Affirm/Magento2_Affirm
# Adds declarative schema and remove installation scripts

---
Deprecate Install/Upgrade/Uninstall Scripts in favor of Declarative Schema
---

### Overview
According to the [docs](https://developer.adobe.com/commerce/php/development/components/declarative-schema/) the scripts that were responsible for installing, upgrading and uninstalling extension modules are now being phased out in favor of declarative schema. This PR captures the required changes for Affirm extension to use declarative schema for installation, upgrade as well as uninstallation processes.  

### Description
The changes are in two parts: declarative schema configuration for Affirm extension module (replaces `InstallSchema` class), and migrating the install, upgrade and uninstall scripts to data patch.

#### db_schema.xml
The schema `etc/db_schema.xml` now replaces the `Setup/InstallSchema.php` when running the `setup:install` or `setup:upgrade` command. According to the above mentioned doc: 

> The new declarative schema approach allows developers to declare the final desired state of the database and has the system adjust to it automatically, without performing redundant operations. Developers are no longer forced to write scripts for each new version. In addition, this approach allows data be deleted when a module is uninstalled.

The specific details on the configuration of the schema can be found [here](https://developer.adobe.com/commerce/php/development/components/declarative-schema/configuration/). 

The schema is responsible for creating two tables, `astound_affirm_rule` and `astound_affirm_attribute` as follows (which maintains the same database schemas as the ones created via scripts):
```
describe astound_affirm_rule;
+-----------------------+--------------+------+-----+---------+----------------+
| Field                 | Type         | Null | Key | Default | Extra          |
+-----------------------+--------------+------+-----+---------+----------------+
| rule_id               | int unsigned | NO   | PRI | NULL    | auto_increment |
| for_admin             | tinyint(1)   | NO   |     | 0       |                |
| is_active             | tinyint(1)   | NO   |     | 0       |                |
| all_stores            | tinyint(1)   | NO   |     | 0       |                |
| all_groups            | tinyint(1)   | NO   |     | 0       |                |
| name                  | varchar(255) | NO   |     | NULL    |                |
| stores                | varchar(255) | NO   |     | NULL    |                |
| cust_groups           | varchar(255) | NO   |     | NULL    |                |
| message               | text         | NO   |     | NULL    |                |
| methods               | text         | YES  |     | NULL    |                |
| conditions_serialized | text         | YES  |     | NULL    |                |
+-----------------------+--------------+------+-----+---------+----------------+

describe astound_affirm_attribute;
+---------+--------------+------+-----+---------+----------------+
| Field   | Type         | Null | Key | Default | Extra          |
+---------+--------------+------+-----+---------+----------------+
| attr_id | int unsigned | NO   | PRI | NULL    | auto_increment |
| rule_id | int unsigned | NO   | MUL | NULL    |                |
| code    | varchar(255) | YES  |     | NULL    |                |
+---------+--------------+------+-----+---------+----------------+

```

#### db_schema_whitelist.json 
Running the install command with declarative schema does not automatically perform db operations that are not defined the `db_schema.xml` and the table, columns or keys must be included in the `etc/db_schema_whitelist.json`. More details can be found [here](https://developer.adobe.com/commerce/php/development/components/declarative-schema/migration-scripts/#create-a-schema-whitelist) .

#### Data patch
A data patch is a class that contains data change instructions, which includes the EAV attributes updates. Currently, only the Product and Category entities are associated with EAV attributes for Affirm and the patch has been implemented in `Setup/Patch/Data/AddProductCategoryAttributes.php` to replace the scripts:
```
affirm_customer_mfp            | Multiple Financing Program value                  
affirm_product_mfp             | Multiple Financing Program value                  
affirm_category_mfp            | Multiple Financing Program value                  
affirm_product_mfp_type        | Multiple Financing Program type                   
affirm_product_mfp_priority    | Multiple Financing Program priority               
affirm_category_mfp_type       | Multiple Financing Program type                   
affirm_category_mfp_priority   | Multiple Financing Program priority               
affirm_product_mfp_start_date  | Start date for time based Financing Program value 
affirm_product_mfp_end_date    | End date for time based Financing Program value   
affirm_category_mfp_start_date | Start date for time based Financing Program value 
affirm_category_mfp_end_date   | End date for time based Financing Program value   
affirm_category_promo_id       | Affirm Promo ID                                   
affirm_product_promo_id        | Affirm Promo ID
```
Note that EAV attributes associated with the Customer Entity have been deprecated and removed from this data patch. 

### How To Test
Steps to test:
1. Clone this repository and checkout to this PR branch.
2. To install, run `bin/magento setup:upgrade`
3. Check that the tables and attributes are correctly populated in the database (`SELECT * FROM eav_attribute WHERE attribute_code LIKE '%affirm%'; and `SHOW TABLES LIKE '%affirm%';`)
4. To uninstall, run `bin/magento module:uninstall --remove-data Astound_Affirm`
